lua/neo-tree/sources/filesystem/lib/fs_actions.lua

```diff
@@ -482,6 +482,113 @@ M.delete_nodes = function(paths_to_delete, callback)
   end)
 end
 
+-- Trash Node
+M.trash_node = function(path, callback, noconfirm)
+  local _, name = utils.split_path(path)
+  local msg = string.format("Are you sure you want to trash '%s'?", name)
+
+  log.trace("Trashing node: ", path)
+  local _type = "unknown"
+  local stat = loop.fs_stat(path)
+  if stat then
+    _type = stat.type
+    if _type == "link" then
+      local link_to = loop.fs_readlink(path)
+      if not link_to then
+        log.error("Could not read link")
+        return
+      end
+      _type = loop.fs_stat(link_to)
+    end
+    if _type == "directory" then
+      local children = scan.scan_dir(path, {
+        hidden = true,
+        respect_gitignore = false,
+        add_dirs = true,
+        depth = 1,
+      })
+      if #children > 0 then
+        msg = "WARNING: Dir not empty! " .. msg
+      end
+    end
+  else
+    log.warn("Could not read file/dir:", path, stat, ", attempting to trash anyway...")
+    -- Guess the type by whether it appears to have an extension
+    if path:match("%.(.+)$") then
+      _type = "file"
+    else
+      _type = "directory"
+    end
+    return
+  end
+
+  local do_trash = function(confirmed)
+    if not confirmed then
+      return
+    end
+
+    if _type == "directory" then
+      local success = false
+      if utils.is_windows then
+        -- TODO: implement trashing for windows
+        log.warn("Trashing is not supported on windows")
+      else
+        local result = vim.fn.system({ "gio", "trash", path })
```

Comment:
Trash command should be configurable. `gio` is probably only available on systems with GTK. MacOS has `trash` command. Windows probably has someting.

I would like to use this wrapper script with `neo-tree`: https://github.com/MunifTanjim/dotfiles/blob/f5c1ae9be297199a1188df9218ba2f921bf4af26/private_dot_config/lf/executable_trash

Reply:
Actually `gio` is available on every system with `glib2` which is almost every UNIX system.

IIRC MacOS doesn't come with `trash` and needs to be installed.

Yes windows does come with `Recycle`

If you want to use an external trashing utility (such as your script) you could put it somewhere on your `$PATH` and pass its name into the trash config this PR would provide.

Reply:
I wonder, does there need to be a separation between "delete" and "trash"? Maybe it can be a single `delete` command that has a config option for what command is executed. Lets call that config option `command`. If that command is `nil`, we stick with the current OS specific logic. If it is a list, we append the path to that list and send it to `vim.fn.system` as-is.

Reply:
An interesting approach.
Speaking of which how do I read configs?
Could you point a place in the codebase that's doing something based on configs.

Reply:
Check out the `add` command:

```lua
      ["a"] = {
        "add",
        -- some commands may take optional config options, see `:h neo-tree-mappings` for details
        config = {
          show_path = "none" -- "none", "relative", "absolute"
        }
      },
```
That config object will be added to the `state` object that is passed to the command function, you can see it being used here:
https://github.com/nvim-neo-tree/neo-tree.nvim/blob/a922c3aff32f3a9498466dd1aa99e14df360fbc5/lua/neo-tree/sources/common/commands.lua#L31